File: cpp/d3q27_bgk_lid.cpp
#include <arrayfire.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// WORK IN PROGRESS !!!

using namespace af;

Window *win;

array mul(const array &a, const array &b) { return a * b; }

array normalize(array a)
{
  return (a / (max<float>(abs(a)) * 1.1)) + 0.1;
}

array stream(array f)
{
  // nearest-neighbours
  f(span, span, span, 1) = shift(f, 1, 0, 0)(span, span, span, 1);
  f(span, span, span, 2) = shift(f, -1, 0, 0)(span, span, span, 2);
  f(span, span, span, 3) = shift(f, 0, 1, 0)(span, span, span, 3);
  f(span, span, span, 4) = shift(f, 0, -1, 0)(span, span, span, 4);
  f(span, span, span, 5) = shift(f, 0, 0, 1)(span, span, span, 5);
  f(span, span, span, 6) = shift(f, 0, 0, -1)(span, span, span, 6);
  // next-nearest neighbours
  // xy plane
  f(span, span, span, 7) = shift(f, 1, 1, 0)(span, span, span, 7);
  f(span, span, span, 8) = shift(f, -1, 1, 0)(span, span, span, 8);
  f(span, span, span, 9) = shift(f, 1, -1, 0)(span, span, span, 9);
  f(span, span, span, 10) = shift(f, -1, -1, 0)(span, span, span, 10);
  // xz plane
  f(span, span, span, 11) = shift(f, 1, 0, 1)(span, span, span, 11);
  f(span, span, span, 12) = shift(f, -1, 0, 1)(span, span, span, 12);
  f(span, span, span, 13) = shift(f, 1, 0, -1)(span, span, span, 13);
  f(span, span, span, 14) = shift(f, -1, 0, -1)(span, span, span, 14);
  // yz plane
  f(span, span, span, 15) = shift(f, 0, 1, 1)(span, span, span, 15);
  f(span, span, span, 16) = shift(f, 0, -1, 1)(span, span, span, 16);
  f(span, span, span, 17) = shift(f, 0, 1, -1)(span, span, span, 17);
  f(span, span, span, 18) = shift(f, 0, -1, -1)(span, span, span, 18);
  // next next-nearest neighbours
  f(span, span, span, 19) = shift(f, 1, 1, 1)(span, span, span, 19);
  f(span, span, span, 20) = shift(f, -1, 1, 1)(span, span, span, 20);
  f(span, span, span, 21) = shift(f, 1, -1, 1)(span, span, span, 21);
  f(span, span, span, 22) = shift(f, -1, -1, 1)(span, span, span, 22);
  f(span, span, span, 23) = shift(f, 1, 1, -1)(span, span, span, 23);
  f(span, span, span, 24) = shift(f, -1, 1, -1)(span, span, span, 24);
  f(span, span, span, 25) = shift(f, 1, -1, -1)(span, span, span, 25);
  f(span, span, span, 26) = shift(f, -1, -1, -1)(span, span, span, 26);
  return f;
}

static void lbm(bool console)
{
  // Grid length, number and spacing
  const unsigned nx = 100;
  const unsigned ny = 100;
  const unsigned nz = 100;

  const unsigned total_nodes = nx * ny * nz;

  // Physical parameters.
  const float ux_lid = 0.1; // horizontal lid velocity
  const float uy_lid = 0;
  const float uz_lid = 0;
  const float rho0 = 1.0;
  // Reynolds number
  float Re = 150.0;
  // Kinematic viscosity
  float nu = ux_lid * nx / Re;
  // Relaxation time
  float tau = 3 * nu + 0.5;
  // Relaxation parameter
  float omega = 1.0 / tau;

  printf("Horizontal lid velocity ux_lid: %f\n", ux_lid);
  printf("Reynolds number: %f\n", Re);
  printf("Lattice viscosity: %f\n", nu);
  printf("Relaxation time: %f\n", tau);
  printf("Relaxation parameter: %f\n", omega);

  const float t1 = 8. / 27.;
  const float t2 = 2. / 27.;
  const float t3 = 1. / 54.;
  const float t4 = 1. / 216.;

  array x = tile(range(nx), 1, ny);
  array y = tile(range(dim4(1, ny), 1), nx, 1);
  array z = tile(range(dim4(1, nz), 1), nx * ny);
  seq lidx = seq(1,nx-2);
  seq lidy = seq(1,ny-2);

  // Discrete velocities
  int cx[27] = {0, 1,-1, 0, 0, 0, 0, 1,-1, 1,-1, 1,-1, 1,-1, 0, 0, 0, 0, 1,-1, 1,-1, 1,-1, 1,-1};
  int cy[27] = {0, 0, 0, 1,-1, 0, 0, 1, 1,-1,-1, 0, 0, 0, 0, 1,-1, 1,-1, 1, 1,-1,-1, 1, 1,-1,-1};
  int cz[27] = {0, 0, 0, 0, 0, 1,-1, 0, 0, 0, 0, 1, 1,-1,-1, 1, 1,-1,-1, 1, 1, 1, 1,-1,-1,-1,-1};
  array ex(27, cx);
  array ey(27, cy);
  array ez(27, cz);

  // weights
  float weights[27] = {t1,t2,t2,t2,t2,t2,t2,t3,t3,t3,t3,t3,t3,t3,t3,t3,t3,t3,t3,t4,t4,t4,t4,t4,t4,t4,t4};
  array w(27, weights);

  array CI = (range(dim4(1, 26), 1) + 1) * total_nodes;
                         // 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26
  unsigned int nb_index_arr[26] = {1,0,3,2,5,4,9,8,7, 6,13,12,11,10,17,16,15,14,25,24,23,22,21,20,19,18};
  array nbidx(26, nb_index_arr);
  array NBI = CI(span, nbidx);

  array main_index_4d = moddims(range(dim4(total_nodes*27)),nx,ny,nz,27);
  array nb_index = flat(stream(main_index_4d));

  // Open lid
  array BOUND = constant(1, nx, ny, nz);
  BOUND(lidx,lidy,seq(1,end)) = 0; // all empty except top lid

  // matrix offset of each Occupied Node
  array ON = where(BOUND);

  // Bounceback indexes
  array TO_REFLECT = flat(tile(ON,CI.elements())) + flat(tile(CI,ON.elements()));
  array REFLECTED = flat(tile(ON,NBI.elements())) + flat(tile(NBI,ON.elements()));

  array DENSITY = constant(rho0, nx, ny, nz);
  array UX = constant(0, nx, ny, nz);
  array UY = constant(0, nx, ny, nz);
  array UZ = constant(0, nx, ny, nz);

  array w_tiled = flat(tile(transpose(w), total_nodes));
  array ex_tiled = flat(tile(transpose(ex), total_nodes));
  array ey_tiled = flat(tile(transpose(ey), total_nodes));
  array ez_tiled = flat(tile(transpose(ez), total_nodes));

  // Particle distribution function in initial equilibrium state
  array u_sq = flat(pow(UX, 2) + af::pow(UY, 2) + af::pow(UZ, 2));
  array eu = flat(batchFunc(transpose(ex), flat(UX), mul) + batchFunc(transpose(ey), flat(UY), mul) + batchFunc(transpose(ez), flat(UZ), mul));
  array F = flat(batchFunc(transpose(w), flat(DENSITY), mul)) * (1.0f + 3.0f*eu + 4.5f*(pow(eu,2)) - 1.5f*(tile(u_sq,27)));

  array uu = constant(0,nx,ny,nz);


  // Setup Window
  win = new Window(1500, 1500, "LBM solver using ArrayFire");
  win->grid(2, 2);

  unsigned iter = 0;
  unsigned maxiter = 15000;

  sync();
  timer::start();

  while (!win->close() && iter < maxiter)
  {
    // Streaming by reading from neighbors (with pre-built index) - pull scheme
    array F_streamed = F(nb_index);

    array BOUNCEDBACK = F_streamed(TO_REFLECT); // Densities bouncing back at next timestep

    array F_2D = moddims(F_streamed, total_nodes, 27);

    // Compute macroscopic variables
    array rho = sum(F_2D, 1);
    DENSITY = moddims(rho,nx,ny,nz);

    array fex = batchFunc(transpose(ex), F_2D, mul);
    array fey = batchFunc(transpose(ey), F_2D, mul);
    array fez = batchFunc(transpose(ez), F_2D, mul);

    UX = moddims((sum(fex, 1) / rho),nx,ny,nz);
    UY = moddims((sum(fey, 1) / rho),nx,ny,nz);
    UZ = moddims((sum(fez, 1) / rho),nx,ny,nz);

    // Macroscopic (Dirichlet) boundary conditions
    UX(lidx,lidy,end) = ux_lid; // lid x - velocity
    UY(lidx,lidy,end) = uy_lid; // lid y - velocity

    UX(ON) = 0;
    UY(ON) = 0;
    UZ(ON) = 0;
    DENSITY(ON) = 0;

    // Collision
    u_sq = flat(pow(UX, 2) + af::pow(UY, 2) + af::pow(UZ, 2));
    eu = flat(batchFunc(transpose(ex), flat(UX), mul) + batchFunc(transpose(ey), flat(UY), mul) + batchFunc(transpose(ez), flat(UZ), mul));
    array FEQ = flat(batchFunc(transpose(w), flat(DENSITY), mul)) * (1.0f + 3.0f*eu + 4.5f*(pow(eu,2)) - 1.5f*(tile(u_sq,27)));

    F = omega * FEQ + (1 - omega) * F_streamed;

    F(REFLECTED) = BOUNCEDBACK;


    if (iter % 10 == 0) {
      uu = moddims(sqrt(u_sq),nx,ny,nz);
      uu(ON) = af::NaN;

      seq filter =  seq(0,nx-1,(int)ceil(nx/30));

      const char *str = "Velocity field for iteration ";
      std::stringstream title;
      title << str << iter;
      (*win)(0, 0).setColorMap(AF_COLORMAP_SPECTRUM);
      (*win)(0, 0).image(flip(transpose(reorder(normalize(uu), 1, 0, 2)(span, span, (int)ceil(nz / 2))), 0));
      (*win)(0, 1).setAxesLimits(0.0f,(float)nx,0.0f,(float)ny,true);
      (*win)(0, 1).vectorField(flat(x(filter,filter)), flat(y(filter,filter)), flat(UX(filter,filter,(int)ceil(nz / 2))), flat(UY(filter,filter,(int)ceil(nz / 2))), std::move(title).str().c_str());
      (*win)(1, 0).setColorMap(AF_COLORMAP_SPECTRUM);
      (*win)(1, 0).image(normalize(uu)(span, span, (int)ceil(nz / 2)));
      (*win)(1, 1).setAxesLimits(0.0f,(float)nx,0.0f,(float)nz,true);
      (*win)(1, 1).vectorField(flat(x(filter,filter)), flat(z(filter,filter)), flat(UX(filter,filter,(int)ceil(nz / 2))), flat(UZ(filter,filter,(int)ceil(nz / 2))), std::move(title).str().c_str());
      win->show();
    }

    if (iter % 100 == 0) {
      float time = timer::stop();
      float mlups = (total_nodes * iter * 10e-6) / time;
      printf("%u iterations completed, %fs elapsed (%f MLUPS).\n", iter, time, mlups);
    }

    iter++;
  }

  sync();

  float end = timer::stop();
  float mlups = (total_nodes * iter * 10e-6) / end;
  printf("Iterations: %d\n", iter);
  printf("Time: %fs\n", end);
  printf("MLUPS: %f\n", mlups);
  af::info();

  while (!win->close())
  {
    uu = sqrt(moddims(u_sq,nx,ny,nz));
    uu(ON) = af::NaN;

    seq filter = seq(0,nx-1,(int)ceil(nx/30));

    const char *str = "Velocity field for iteration ";
    std::stringstream title;
    title << str << iter;
    (*win)(0, 0).setColorMap(AF_COLORMAP_SPECTRUM);
    (*win)(0, 0).image(flip(transpose(normalize(uu)),0));
    (*win)(0, 1).setAxesLimits(0.0f,(float)nx,0.0f,(float)ny,true);
    (*win)(0, 1).vectorField(flat(x(filter,filter)), flat(y(filter,filter)), flat(UX(filter,filter,(int)ceil(nz / 2))), flat(UY(filter,filter,(int)ceil(nz / 2))), std::move(title).str().c_str());
    win->show();
  }
}

int main(int argc, char *argv[])
{
  int device = argc > 1 ? atoi(argv[1]) : 0;
  bool console = argc > 2 ? argv[2][0] == '-' : false;
  try
  {
    af::setDevice(device);
    af::info();
    printf("LBM D2Q9 simulation\n");
    lbm(console);
  }
  catch (exception &e)
  {
    fprintf(stderr, "%s\n", e.what());
    throw;
  }

  return 0;
}

File: rust/examples/d3q27_bgk_lid.rs
use arrayfire::*;
use std::time::Instant;
use std::error::Error;
use csv::Writer;

type FloatNum = f32;

fn normalize(a: &Array<FloatNum>) -> Array<FloatNum> {
    (a / (max_all(&abs(a)).0 * 1.1 as FloatNum)) + 0.1 as FloatNum
}

fn stream(f: &Array<FloatNum>) -> Array<FloatNum> {
    let mut pdf = f.clone();
    // nearest-neighbours
    eval!(pdf[1:1:0, 1:1:0, 1:1:1] = shift(&view!(f[1:1:0, 1:1:0, 1:1:1]), &[ 1, 0, 0, 0]));
    eval!(pdf[1:1:0, 1:1:0, 2:2:1] = shift(&view!(f[1:1:0, 1:1:0, 2:2:1]), &[ -1, 0, 0, 0]));
    eval!(pdf[1:1:0, 1:1:0, 3:3:1] = shift(&view!(f[1:1:0, 1:1:0, 3:3:1]), &[ 0, 1, 0, 0]));
    eval!(pdf[1:1:0, 1:1:0, 4:4:1] = shift(&view!(f[1:1:0, 1:1:0, 4:4:1]), &[ 0, -1, 0, 0]));
    eval!(pdf[1:1:0, 1:1:0, 5:5:1] = shift(&view!(f[1:1:0, 1:1:0, 5:5:1]), &[ 0, 0, 1, 0]));
    eval!(pdf[1:1:0, 1:1:0, 6:6:1] = shift(&view!(f[1:1:0, 1:1:0, 6:6:1]), &[ 0, 0, -1, 0]));
    // next-nearest neighbours
    // xy plane
    eval!(pdf[1:1:0, 1:1:0, 7:7:1] = shift(&view!(f[1:1:0, 1:1:0, 7:7:1]), &[ 1, 1, 0, 0]));
    eval!(pdf[1:1:0, 1:1:0, 8:8:1] = shift(&view!(f[1:1:0, 1:1:0, 8:8:1]), &[ -1, 1, 0, 0]));
    eval!(pdf[1:1:0, 1:1:0, 9:9:1] = shift(&view!(f[1:1:0, 1:1:0, 9:9:1]), &[ 1, -1, 0, 0]));
    eval!(pdf[1:1:0, 1:1:0, 10:10:1] = shift(&view!(f[1:1:0, 1:1:0, 10:10:1]), &[ -1, -1, 0, 0]));
    // xz plane
    eval!(pdf[1:1:0, 1:1:0, 11:11:1] = shift(&view!(f[1:1:0, 1:1:0, 11:11:1]), &[ 1, 0, 1, 0]));
    eval!(pdf[1:1:0, 1:1:0, 12:12:1] = shift(&view!(f[1:1:0, 1:1:0, 12:12:1]), &[ -1, 0, 1, 0]));
    eval!(pdf[1:1:0, 1:1:0, 13:13:1] = shift(&view!(f[1:1:0, 1:1:0, 13:13:1]), &[ 1, 0, -1, 0]));
    eval!(pdf[1:1:0, 1:1:0, 14:14:1] = shift(&view!(f[1:1:0, 1:1:0, 14:14:1]), &[ -1, 0, -1, 0]));
    // yz plane
    eval!(pdf[1:1:0, 1:1:0, 15:15:1] = shift(&view!(f[1:1:0, 1:1:0, 15:15:1]), &[ 0, 1, 1, 0]));
    eval!(pdf[1:1:0, 1:1:0, 16:16:1] = shift(&view!(f[1:1:0, 1:1:0, 16:16:1]), &[ 0, -1, 1, 0]));
    eval!(pdf[1:1:0, 1:1:0, 17:17:1] = shift(&view!(f[1:1:0, 1:1:0, 17:17:1]), &[ 0, 1, -1, 0]));
    eval!(pdf[1:1:0, 1:1:0, 18:18:1] = shift(&view!(f[1:1:0, 1:1:0, 18:18:1]), &[ 0, -1, -1, 0]));
    // next next-nearest neighbours
    eval!(pdf[1:1:0, 1:1:0, 19:19:1] = shift(&view!(f[1:1:0, 1:1:0, 19:19:1]), &[ 1, 1, 1, 0]));
    eval!(pdf[1:1:0, 1:1:0, 20:20:1] = shift(&view!(f[1:1:0, 1:1:0, 20:20:1]), &[ -1, 1, 1, 0]));
    eval!(pdf[1:1:0, 1:1:0, 21:21:1] = shift(&view!(f[1:1:0, 1:1:0, 21:21:1]), &[ 1, -1, 1, 0]));
    eval!(pdf[1:1:0, 1:1:0, 22:22:1] = shift(&view!(f[1:1:0, 1:1:0, 22:22:1]), &[ -1, -1, 1, 0]));
    eval!(pdf[1:1:0, 1:1:0, 23:23:1] = shift(&view!(f[1:1:0, 1:1:0, 23:23:1]), &[ 1, 1, -1, 0]));
    eval!(pdf[1:1:0, 1:1:0, 24:24:1] = shift(&view!(f[1:1:0, 1:1:0, 24:24:1]), &[ -1, 1, -1, 0]));
    eval!(pdf[1:1:0, 1:1:0, 25:25:1] = shift(&view!(f[1:1:0, 1:1:0, 25:25:1]), &[ 1, -1, -1, 0]));
    eval!(pdf[1:1:0, 1:1:0, 26:26:1] = shift(&view!(f[1:1:0, 1:1:0, 26:26:1]), &[ -1, -1, -1, 0]));

    pdf
}

fn output_csv(mlups: Vec<f32>) -> Result<(), Box<dyn Error>> {
  let mut wtr = Writer::from_path("d3q27_bgk_lid_mlups.csv")?;

  wtr.write_record(&["Iterations", "MLUPS"])?;
  for (i, item) in mlups.iter().enumerate() {
    wtr.write_record(&[i.to_string(), item.to_string()])?;
  }

  wtr.flush()?;
  Ok(())
}

fn lbm(write_csv: bool) {
    // Grid length, number and spacing
    let nx: u64 = 100;
    let ny: u64 = 100;
    let nz: u64 = 100;

    let total_nodes = nx * ny * nz;

    // Physical parameters.
    let ux_lid: FloatNum = 0.05; // horizontal lid velocity
    let uy_lid: FloatNum = 0.0; // vertical lid velocity
    let uz_lid: FloatNum = 0.0; // lid velocity in z-direction
    let rho0: FloatNum = 1.0;

    // Reynolds number
    let re: FloatNum = 100.0;
    // Kinematic viscosity
    let nu: FloatNum = ux_lid * nx as FloatNum / re;
    // Relaxation time
    let tau: FloatNum = (3.0 as FloatNum) * nu + (0.5 as FloatNum);
    // Relaxation parameter
    let omega: FloatNum = (1.0 as FloatNum) / tau;

    println!("Horizontal lid velocity ux_lid: {}", ux_lid);
    println!("Vertical lid velocity uy_lid: {}", uy_lid);
    println!("Reynolds number: {}", re);
    println!("Lattice viscosity: {}", nu);
    println!("Relaxation time: {}", tau);
    println!("Relaxation parameter: {}", omega);

    let t1: FloatNum = 8. / 27.;
    let t2: FloatNum = 2. / 27.;
    let t3: FloatNum = 1. / 54.;
    let t4: FloatNum = 1. / 216.;

    let x: Array<FloatNum> = tile(&range(dim4!(nx), 1), dim4!(ny * nz));
    let y: Array<FloatNum> = tile(&range(dim4!(1, ny), 1), dim4!(nx, nz));
    let z: Array<FloatNum> = tile(&range(dim4!(1, nz), 1), dim4!(nx, ny));

    let dims = dim4!(nx, ny, nz);
    let span = seq!();

    let ux_lid_af = constant::<FloatNum>(ux_lid, dims);
    let uy_lid_af = constant::<FloatNum>(uy_lid, dims);
    let uz_lid_af = constant::<FloatNum>(uz_lid, dims);

    let lidx = seq!(1, nx as i32 - 2, 1);
    let lidy = seq!(1, ny as i32 - 2, 1);
    let end_y = seq!(nx as i32 - 1, ny as i32 - 1, 1);

    // Discrete velocities
    let ex = Array::<FloatNum>::new(&[0., 1., 0., -1., 0., 1., -1., -1., 1.], dim4!(27));
    let ey = Array::<FloatNum>::new(&[0., 0., 1., 0., -1., 1., 1., -1., -1.], dim4!(27));
    let ez = Array::<FloatNum>::new(&[0., 0., 1., 0., -1., 1., 1., -1., -1.], dim4!(27));

    // weights
    let w = Array::new(&[t1,t2,t2,t2,t2,t2,t2,t3,t3,t3,t3,t3,t3,t3,t3,t3,t3,t3,t3,t4,t4,t4,t4,t4,t4,t4,t4], dim4!(27));

    let ci: Array<u64> = (range::<u64>(dim4!(1, 26), 1) + 1) * total_nodes;
    let nbidx = Array::new(&[1,0,3,2,5,4,9,8,7,6,13,12,11,10,17,16,15,14,25,24,23,22,21,20,19,18], dim4!(26));
    let nbi: Array<u64> = view!(ci[span, nbidx]);

    let main_index = moddims(&range(dim4!(total_nodes * 27), 0), dim4!(nx, ny, nz, 27));
    let nb_index = flat(&stream(&main_index));

    // Open lid 3D
    let mut bound = constant::<FloatNum>(1.0, dims);
    let zeros = constant::<FloatNum>(0.0, dims);
    let all_except_top_lid = seq!(1, nz as i32 - 1, 1);
    assign_seq(
        &mut bound,
        &[lidx, lidy, all_except_top_lid],
        &index(&zeros, &[lidx, lidy, all_except_top_lid]),
    );

    // matrix offset of each Occupied Node
    let on = locate(&bound);

    // Bounceback indexes
    let to_reflect = flat(&tile(&on, dim4!(ci.elements() as u64)))
        + flat(&tile(&ci, dim4!(on.elements() as u64)));
    let reflected = flat(&tile(&on, dim4!(nbi.elements() as u64)))
        + flat(&tile(&nbi, dim4!(on.elements() as u64)));

    let mut density = constant::<FloatNum>(rho0, dims);
    let mut ux = constant::<FloatNum>(0.0, dims);
    let mut uy = constant::<FloatNum>(0.0, dims);
    let mut uz = constant::<FloatNum>(0.0, dims);

    let zeroed_on = constant::<FloatNum>(0.0, on.dims());

    // Start in equilibrium state
    let mut u_sq: Array<FloatNum> =
        flat(&(pow(&ux, &(2.0 as FloatNum), false) + pow(&uy, &(2.0 as FloatNum), false) + pow(&uz, &(2.0 as FloatNum), false)));
    let mut eu: Array<FloatNum> = flat(
        &(&mul(&transpose(&ex, false), &flat(&ux), true)
            + &mul(&transpose(&ey, false), &flat(&uy), true)
            + &mul(&transpose(&ez, false), &flat(&uy), true)),
    );
    let mut f: Array<FloatNum> = flat(&mul(&transpose(&w, false), &flat(&density), true))
        * ((1.0 as FloatNum)
            + (3.0 as FloatNum) * &eu
            + (4.5 as FloatNum) * (&pow(&eu, &(2.0 as FloatNum), false))
            - (1.5 as FloatNum) * (&tile(&flat(&u_sq), dim4!(27))));

    // Create a window to show the waves.
    // let mut win = Window::new(1536, 768, "LBM solver using ArrayFire".to_string());
    // win.grid(2, 2);

    let mut iter: u64 = 0;
    let maxiter: u64 = 5000;
    let mut mlups: Vec<FloatNum> = Vec::with_capacity(5000);

    sync(0);
    let timer = Instant::now();

    mem_info!("Before benchmark");

    while iter < maxiter {
        // Streaming by reading from neighbors (with pre-built index) - pull scheme
        let f_streamed = view!(f[nb_index]);

        let bouncedback = view!(f_streamed[to_reflect]); // Densities bouncing back at next timestep

        let f_2d = moddims(&f_streamed, dim4!(total_nodes, 27));

        // Compute macroscopic variables
        let rho = sum(&f_2d, 1);
        density = moddims(&rho, dims);

        let fex = mul(&transpose(&ex, false), &f_2d, true);
        let fey = mul(&transpose(&ey, false), &f_2d, true);
        let fez = mul(&transpose(&ez, false), &f_2d, true);

        ux = moddims(&(sum(&fex, 1) / &rho), dims);
        uy = moddims(&(sum(&fey, 1) / &rho), dims);
        uz = moddims(&(sum(&fez, 1) / &rho), dims);

        // Macroscopic (Dirichlet) boundary conditions
        eval!(ux[lidx, lidy, end_y] = view!(ux_lid_af[lidx, lidy, end_y]));
        eval!(uy[lidx, lidy, end_y] = view!(uy_lid_af[lidx, lidy, end_y]));
        eval!(uz[lidx, lidy, end_y] = view!(uz_lid_af[lidx, lidy, end_y]));

        eval!(ux[on] = zeroed_on);
        eval!(uy[on] = zeroed_on);
        eval!(uz[on] = zeroed_on);
        eval!(density[on] = zeroed_on);

        // Collision
        u_sq = flat(&(pow(&ux, &(2.0 as FloatNum), false) + pow(&uy, &(2.0 as FloatNum), false) + pow(&uz, &(2.0 as FloatNum), false)));
        eu = flat(
            &(&mul(&transpose(&ex, false), &flat(&ux), true)
                + &mul(&transpose(&ey, false), &flat(&uy), true)
                + &mul(&transpose(&ez, false), &flat(&uy), true)),
        );
        let feq = flat(&mul(&transpose(&w, false), &flat(&density), true))
            * ((1.0 as FloatNum)
                + (3.0 as FloatNum) * &eu
                + (4.5 as FloatNum) * (&pow(&eu, &(2.0 as FloatNum), false))
                - (1.5 as FloatNum) * (&tile(&flat(&u_sq), dim4!(27))));

        f = omega * feq + (1.0 - omega) * f_streamed;

        eval!(f[reflected] = bouncedback);

        // Visualization
        // if iter % 10 == 0 {
        //     let mut uu = moddims(&sqrt(&u_sq), dims);
        //     eval!(uu[on] = constant::<FloatNum>(FloatNum::NAN, on.dims()));

        //     let filter = seq!(0, nx as i32 - 1, nx as i32 / 30);
        //     let z_section = seq!(nz as i32 / 2, nz as i32 / 2, 1);

        //     let xy_view = index(&reorder_v2(&normalize(&uu), 1, 0, Some(vec![2])), &[span, span, z_section]);

        //     win.set_view(0, 0);
        //     win.set_colormap(ColorMap::SPECTRUM);
        //     win.draw_image(
        //       &flip(&transpose(&xy_view, false), 0),
        //       Some(format!("XY domain in iteration {}", &iter).to_string()),
        //     );

        //     // win.set_view(0, 1);
        //     // win.set_axes_limits_2d(0.0, nx as f32, 0.0, ny as f32, true);
        //     // win.draw_vector_field2(
        //     //     &flat(&view!(x[filter,filter])),
        //     //     &flat(&view!(y[filter,filter])),
        //     //     &flat(&view!(ux[filter,filter,z_section])),
        //     //     &flat(&view!(uy[filter,filter,z_section])),
        //     //     Some(format!("Velocity field in iteration {}", &iter).to_string()),
        //     // );

        //     win.set_view(1, 0);
        //     win.set_colormap(ColorMap::SPECTRUM);
        //     win.draw_image(
        //       &normalize(&index(&uu, &[span, span, z_section])),
        //       Some(format!("XY domain in iteration {}", &iter).to_string()),
        //     );

        //     win.show();
        // }

        let time = timer.elapsed().as_secs() as FloatNum;
        let updates = (total_nodes as FloatNum * iter as FloatNum * 10e-6) / time;

        if !updates.is_nan() && !updates.is_infinite() {
          mlups.push(updates);
        }

        if iter % 100 == 0 {
            println!(
                "{} iterations completed, {}s elapsed ({} MLUPS).",
                iter, time, updates
            );
        }

        iter += 1;
        sync(0);
    }

    mem_info!("After benchmark");
    sync(0);

    // output CSV of MLUPS data
    if write_csv {
      output_csv(mlups);
    }
}

fn main() {
    set_device(0);
    set_backend(Backend::OPENCL);
    info();
    println!("LBM D2Q9 simulation\n");
    let write_csv = true;
    lbm(write_csv);
}
